### PR TITLE
MRG: Fix RtEpochs to support saving

### LINKED ISSUE #5191: Saving RtEpochs
TL;DR RtEpochs do not support all of the inherited ``BaseEpochs`` methods. It would be nice to have an efficient conversion-method to an ``EpochsArray``.

While ``RtEpochs`` are useful for gathering and processing individual  epochs, it would be helpful to be able to make them completely compatible with an "offline" ``Epochs`` object. For example,   currently it is not possible to save a ``RtEpochs`` object due to  (at least) the following issues:
     - missing parameters (``out`` and ``verbose``) in ``_get_data``
     - ``drop_log`` is missing or incorrect
     
As a workaround, I exported the data to an EpochsArray [1], but noticed that data needs to be copied multiple times:
  1. a single numpy array into which all ``_epoch_queue`` items are copied
  2. the Epochs constructor again copies the data in its call to ``decimate`` (which does not make sense to me anyway, since I would expect a ``decimate(1)`` call to not have any side-effects)

Having three (temporary) copies of the epochs of a full experiment in memory leads frequently to out-of-memory situations on typical workstations.

I would propose to 
  - remove the mandatory call to ``decimate`` or omit the copying when creating data from an existing array (although there is probably a reason the call is currently implemented?)
  - either ensure that ``RtEpochs`` *is a* ``BaseEpochs`` object supporting all the methods or alternatively provide a conversion method to an ``EpochsArray`` etc.

It would be great to hear your thoughts on this. I'm willing to contribute.

### PR DESCRIPTION
**Pull request by hekolk**
Fixes #5191: Make sure the epochs invariants hold  for ``RtEpochs``.

I'm not completely happy with the ``_get_data()`` method since it relies on the "online iteration and blocking" over epochs rather than just returning the data received so far (which I would prefer, but didn't implement for compability).  I had to move the check for the timeout since otherwise iteration is typically only possible once (depending on the ``isi_max``).
The inherited implementation of _get_item was broken (but is called by ``save``).

### CODE REVIEW

File: mne/realtime/epochs.py [hunk 1]
```diff
@@ -208,6 +211,76 @@ def events(self):
...
+        """Update the internal selection list"""
+        self._selection = list(new_selection)
+
+    def __len__(self):
+        """Return the number of available epochs"""
+        # we override this since the check for dropped epochs doesn't apply here
+        return len(self._events)
+
+    def _getitem(self, item, reason='IGNORED', copy=True, drop_event_id=True):
+        """Wrap __getitem__ with more options."""
```
Comment (larsoner) [inline mne/realtime/epochs.py]: Is there a way not to copy the code? Maybe by adding an extra option to the existing `__getitem__` or calling `super` here?

**Comment by larsoner** [comment]
I agree that `RtEpochs.save(...)` should save immediately without waiting for more epochs

**Comment by hekolk** [comment]
I removed the blocking from ``RtEpochs.get_data()`` (and hence ``RtEpochs.save()``). I hope the additional note in the docstring as well as a "whatsnew"-entry are reasonable.

When adding a test that the new ``drop_log`` and ``selection`` fields work as expected, I fixed a minor bug in ``RtEpochs._process_raw_buffer`` that occurred when buffer size is smaller than epoch size: The update of ``_first_samp`` was missing in case the buffer was not yet at maximum ( https://github.com/mne-tools/mne-python/pull/5193/files#diff-e385d76eea56e4905e874f408888e84eL388 ), causing incorrect checks whether an event can be processed. There probably wasn't a real-life impact if ``RtEpochs`` were started early enough before the first event.

File: mne/realtime/epochs.py [hunk 2]
```diff
@@ -126,6 +126,14 @@ class RtEpochs(BaseEpochs):
         The events associated with the epochs currently in the queue.
     verbose : bool, str, int, or None
         See above.
+
+    Notes
+    -----
+    Calling `next()` on an `RtEpochs` object (as internally done when iterating
+     over the object) is blocking, i.e., waits for at most `isi_max` seconds
```
Comment (larsoner) [inline mne/realtime/epochs.py]: The inconsistent indentation breaks the rendering:

https://8009-1301584-gh.circle-artifacts.com/0/html/generated/mne.realtime.RtEpochs.html

File: mne/realtime/epochs.py [hunk 3]
```diff
@@ -205,9 +215,82 @@ def __init__(self, client, event_id, tmin, tmax, stim_channel='STI 014',
 
     @property
     def events(self):
-        """The events associated with the epochs currently in the queue."""
+        """
+        Returns the events associated with the epochs currently in the queue.
```
Comment (larsoner) [inline mne/realtime/epochs.py]: This is a `property` (not exposed to the user as a function) so I think the original docstring is cleaner

File: mne/realtime/epochs.py [hunk 4]
```diff
@@ -205,9 +215,82 @@ def __init__(self, client, event_id, tmin, tmax, stim_channel='STI 014',
...
+        array of int, shape (n_events, 3)
+        event array
+        """
+        """"""
         return np.array(self._events)
 
+    @events.setter
+    def events(self, new_events):
+        """
+        Updates the internal event list.
```
Comment (larsoner) [inline mne/realtime/epochs.py]: `Update` (imperative voice)

File: mne/realtime/epochs.py [hunk 5]
```diff
@@ -205,9 +215,82 @@ def __init__(self, client, event_id, tmin, tmax, stim_channel='STI 014',
...
+            new events
+        """
+        self._events = [tuple(new_events[i, :])
+                        for i in range(len(new_events))]
+
+    @property
+    def selection(self):
+        """
+        Returns the current epoch selection.
+        Returns
```
Comment (larsoner) [inline mne/realtime/epochs.py]: This docstring is not formatted properly. Better to just say `"""Array of integers of the current selection."""`

File: mne/realtime/tests/test_mockclient.py [hunk 6]
```diff
@@ -29,16 +31,44 @@ def test_mockclient():
...
 
+    # get_data() should return immediately and not wait for the timeout
+    start_time = time.time()
     rt_data = rt_epochs.get_data()
-
+    retrieval_time = time.time() - start_time
+    assert_true(retrieval_time < isi_max)
     assert_true(rt_data.shape == data.shape)
     assert_array_equal(rt_data, data)
+    assert_true(len(rt_epochs) == len(epochs))
```
Comment (larsoner) [inline mne/realtime/tests/test_mockclient.py]: you can't tell from this file but we have stopped using `nose`-based asserts like `assert_true` and `assert_equal` now that we use `pytest`, instead just do:
```
assert len(rt_epochs) == len(epochs)
```

File: mne/realtime/tests/test_mockclient.py [hunk 7]
```diff
@@ -29,16 +31,44 @@ def test_mockclient():
...
+    # iteration over epochs should block until timeout
+    rt_iter_data = list()
+    start_time = time.time()
+    for cur_epoch in rt_epochs:
+        rt_iter_data.append(cur_epoch)
+    retrieval_time = time.time() - start_time
+    assert_true(retrieval_time >= isi_max)
+    rt_iter_data = np.array(rt_iter_data)
+    assert_true(rt_iter_data.shape == data.shape)
+    assert_array_equal(rt_iter_data, data)
```
Comment (larsoner) [inline mne/realtime/tests/test_mockclient.py]: ... but `numpy.testing` comparators like `assert_array_equal` are fine to keep.

File: mne/realtime/tests/test_mockclient.py [hunk 8]
```diff
@@ -184,4 +214,60 @@ def test_find_events():
...
+    rt_epochs.start()
+    rt_client.send_data(rt_epochs, picks, tmin=0, tmax=raw_tmax,
+                        buffer_size=250)
+
+    assert_true(len(rt_epochs) == len(epochs_to_keep))
+    assert_array_equal(rt_epochs.drop_log, expected_drop_log)
+    assert_array_equal(rt_epochs.selection, epochs_to_keep)
+    rt_data = rt_epochs.get_data()
+    assert_true(rt_data.shape == epochs_data.shape)
+    assert_array_equal(rt_data, epochs_data)
```
Comment (larsoner) [inline mne/realtime/tests/test_mockclient.py]: How long do these new test take relative to the ones in `master`? We have to be careful not to add too much to the testing time, I would say ideally < 1 sec for this functionality but if you need 5-10 sec to actually get something from the FT buffer we can probably live with it (but try to make this as quick as possible).

File: mne/realtime/tests/test_mockclient.py (hunk 6 above)
Comment (hekolk) [reply mne/realtime/tests/test_mockclient.py]: Ok, I'll change all the tests I touched - or should I convert the whole file?
Also, do you prefer ``mne.utils._TempDir`` or the ``pytest`` ``tmpdir`` fixture?

File: mne/realtime/tests/test_mockclient.py [hunk 9]
```diff
@@ -29,16 +31,44 @@ def test_mockclient():
...
     assert_array_equal(rt_data, data)
+    assert_true(len(rt_epochs) == len(epochs))
+
+    # iteration over epochs should block until timeout
+    rt_iter_data = list()
+    start_time = time.time()
+    for cur_epoch in rt_epochs:
+        rt_iter_data.append(cur_epoch)
+    retrieval_time = time.time() - start_time
+    assert_true(retrieval_time >= isi_max)
```
Comment (hekolk) [inline mne/realtime/tests/test_mockclient.py]: I added the timeout check to make sure behavior is consistent with the previous implementation. I think, waiting for ``0.5s`` would be long enough to make sure the delay is not due to a slow build environment. But I could also remove the timing part if it takes too long/seems too fragile to you

**Comment by hekolk** [comment]
@larsoner on my machine, ``pytest mne/realtime/tests/*`` takes ``6.9s``, compared to ``6.8s`` on ``master``.

File: mne/realtime/tests/test_mockclient.py (hunk 6 above)
Comment (larsoner) [reply mne/realtime/tests/test_mockclient.py]: We have been migrating one file at a time as people are motivated to do it, so yes please feel free!

We haven't transitioned to using `tmpdir` anywhere yet, but probably should -- feel free to make that change as well and we'll probably use it as a template elsewhere.

**Comment by larsoner** [comment]
Excellent

**Comment by hekolk** [comment]
Regarding the ``tmpdir`` fixture of ``pytest``: One could omit the deletion of the created file since ``pytest`` keeps only the last three directories [0]: Less verbose code, but a little space requirement on developer machines (but probably negligible compared to the testing data; and CI environments are ephemeral anyway).
[0]: https://docs.pytest.org/en/latest/tmpdir.html#the-default-base-temporary-directory

File: mne/realtime/tests/test_mockclient.py [hunk 10]
```diff
@@ -29,16 +30,46 @@ def test_mockclient():
...
+    export_file = tmpdir.join('test_rt-epo.fif')
+    try:
+        rt_epochs.save(str(export_file))
+
+        loaded_epochs = mne.read_epochs(str(export_file))
+        loaded_data = loaded_epochs.get_data()
+        assert rt_data.shape == loaded_data.shape
+        assert_allclose(loaded_data, rt_data)
+    finally:
+        export_file.remove(ignore_errors=True)
```
Comment (larsoner) [inline mne/realtime/tests/test_mockclient.py]: Can we avoid doing this? Based on what you say the `tmpdir` will eventually go away on its own. We don't want to have to add these all over our code, `_TempDir` does not require it.

**Comment by jasmainak** [comment]
> Anything else that should be tested?

Sorry for coming late to the scene. Can you check if the examples here: https://github.com/mne-tools/mne-python/tree/master/examples/realtime work, specially the ones without the `plot_` prefix

File: mne/realtime/epochs.py [hunk 11]
```diff
@@ -208,6 +218,65 @@ def events(self):
...
+        """
+        Return the number of available epochs.
+
+        -------
+        int
+        number of available epochs
+        """
+        # we override this since the check for dropped epochs
+        # doesn't apply here
+        return len(self._events)
```
Comment (jasmainak) [inline mne/realtime/epochs.py]: the rejection dictionary does not work for `RtEpochs` ?

**Comment by hekolk** [comment]
Regarding the Realtime examples:

(1) ``plot_compute_rt_decoder.py`` produces different results due to a different number of events. This is due to the fix in updating ``self._first_samp`` from 7f6b540. To be sure this is a bugfix and not a regression, I added a stripped-down version of the example as a test, including assertions that the detected events and decoding results are identical to the ones from the "offline" ``Epochs`` version. The test requires sample data and takes ``3s`` on my machine. 

The added test fails on the current master (with and without the fixes in ``MockRtClient`` to also send a single partial buffer with the rest of the data) Errors (due to results being different from the offline ``Epochs``'s events) start from the third buffer's content onwards:
```
E       AssertionError: 
E       Arrays are not equal
E       
E       (shapes (48, 3), (52, 3) mismatch)
E        x: array([[  636,     0,     3],
E              [  742,     0,     1],
E              [ 1056,     0,     3],...
E        y: array([[  636,     0,     3],
E              [  742,     0,     1],
E              [ 1056,     0,     3],...
```

(2) The fieldtrip examples require the patched ``neuromag2ft`` which is not available anymore from ``hut.fi``. @jasmainak if it should be tested, could you make it available somewhere else?

(3) The ``StimServer`` examples failed due to a mismatch of picks: ``FilterEstimator`` picks MEG/EEG channels based on the supplied info, whereas the example additionally picked channels on the MockClient's ``get_data()`` calls (I changed the latter). The ``rt_feedback_client`` example currently requires ``psychopy`` (which depends on a vast amount of dependies including ``wxPython`` which does not compile in a ``virtualenv`` when installed through ``pip``) → wouldn't it make sense to replace that with a simple ``print('left stimulus')`` etc. and have an example that runs out-of-the-box!?

File: mne/realtime/tests/test_mockclient.py [hunk 12]
```diff
@@ -180,8 +210,123 @@ def test_find_events():
...
+            X_rt = ev.data[None, data_picks, :]
+            y_rt = int(ev.comment)  # comment attribute contains the event_id
+        else:
+            X_rt = np.concatenate((X_rt, ev.data[None, data_picks, :]), axis=0)
+            y_rt = np.append(y_rt, int(ev.comment))
+
+    X_offline = epochs_offline.get_data()[:, data_picks, :]
+    y_offline = epochs_offline.events[:, 2]
+    assert_array_equal(X_rt, X_offline)
+    assert_array_equal(y_rt, y_offline)
```
Comment (jasmainak) [inline mne/realtime/tests/test_mockclient.py]: I think this is a great idea! We should probably do this also for `FieldTripClient`...

**Comment by larsoner** [comment]
+1 for removing `psychopy` dependency

**Comment by hekolk** [comment]
Alright, all the examples run on my machine (and ``psychopy`` is gone). 

I also added a test that the ``FieldTripClient`` returns all of the epochs as well as the "correct" data (actually the array values were not all close even for larger ``atol`` and ``rtol`` parameters, so right now I only check that a large fraction (98%) of values of the data array is close - not nice, but I suppose better than not testing anything). The test takes a while (``21s``), so I marked it as ``slow``. It is only executed if the ``neuromag2ft`` binary is available, anyway.

**Comment by larsoner** [comment]
> The test takes a while (21s), so I marked it as slow. It is only executed if the neuromag2ft binary is available, anyway.

Can we make this faster by changing the ISI, epoch length, and/or `raw` duration (e.g., via `.crop`) so it is more like 5 sec (without compromising the test quality)?

File: doc/whats_new.rst [hunk 13]
```diff
@@ -33,6 +33,9 @@ Bug
 API
 ~~~
 
+- Calling :meth:`mne.Epochs.decimate` with ``decim=1`` no longer copies the data by `Henrich Kolkhorst`_
+
+- Removed blocking (waiting for new epochs) in :meth:`mne.realtime.RtEpochs.get_data()` by `Henrich Kolkhorst`_
```
Comment (larsoner) [inline doc/whats_new.rst]: You need to also add a URL for yourself at the bottom

File: examples/realtime/ftclient_rt_average.py [hunk 14]
```diff
@@ -12,7 +12,7 @@
 
 using a modified version of neuromag2ft available at
 
-http://neuro.hut.fi/~mainak/neuromag2ft-2.0.0.zip
+https://staff.washington.edu/larsoner/minimal_cmds.tar.gz
```
Comment (larsoner) [inline examples/realtime/ftclient_rt_average.py]: Another potentially unreliable URL but better than a broken one :)

File: examples/realtime/ftclient_rt_compute_psd.py [hunk 15]
```diff
@@ -41,6 +42,10 @@
 
     n_fft = 256  # the FFT size. Ideally a power of 2
     n_samples = 2048  # time window on which to compute FFT
+
+    # make sure at least one epoch is available
+    time.sleep(float(n_samples) / raw_info['sfreq'])
```
Comment (larsoner) [inline examples/realtime/ftclient_rt_compute_psd.py]: no need for `float` we should always have that `raw_info['sfreq']` is a float (if it's not, we have a bug elsewhere)

File: mne/realtime/epochs.py [hunk 16]
```diff
@@ -126,6 +126,14 @@ class RtEpochs(BaseEpochs):
         The events associated with the epochs currently in the queue.
     verbose : bool, str, int, or None
         See above.
+
+    Notes
+    -----
+    Calling `next()` on an `RtEpochs` object (as internally done when iterating
+    over the object) is blocking, i.e., waits for at most `isi_max` seconds
+    for a new epoch to be received.
+    Calling `get_data()` on an `RtEpochs` object immediately returns the epochs
```
Comment (larsoner) [inline mne/realtime/epochs.py]: Did you want these to be separate paragraphs? They won't be unless you add an extra newline.

Maybe they should be bullets (created by adding `* ` or `- ` before start lines and indenting subsequent lines)?

**Comment by jasmainak** [comment]
> Can we make this faster by changing the ISI, epoch length, and/or raw duration (e.g., via .crop) so it is more like 5 sec (without compromising the test quality)?

There should be an option in `neuromag2ft` called `--speed` to make the simulation faster. Does that help?

File: mne/realtime/tests/test_fieldtrip_client.py [hunk 17]
```diff
@@ -22,33 +24,95 @@
...
+                    events_ids_rt = np.append(events_ids_rt, int(ev.comment))
+
+            rt_epochs.stop()
+
+        assert_array_equal(events_ids_rt, rt_epochs.events[:, 2])
+        assert_array_equal(X_rt, rt_epochs.get_data())
+        assert len(rt_epochs) == len(epochs_offline)
+        assert_array_equal(events_ids_rt, epochs_offline.events[:, 2])
+        # we don't get exactly the same values (type conversions?)
+        # check only that "almost all" values are close
```
Comment (jasmainak) [inline mne/realtime/tests/test_fieldtrip_client.py]: that's weird. Are you sure it's not a missing epoch?

File: mne/realtime/tests/test_fieldtrip_client.py [hunk 18]
```diff
@@ -22,33 +24,95 @@
...
+    tmin, tmax = -0.2, 0.5
+    epochs_offline = mne.Epochs(raw, events_offline, event_id=event_id,
+                                tmin=tmin, tmax=tmax)
+    epochs_offline.drop_bad()
+    isi_max = (np.max(np.diff(epochs_offline.events[:, 0])) /
+               float(raw.info['sfreq'])) + 7.0
+
+    kill_signal = _start_buffer_thread()
+
+    try:
```
Comment (jasmainak) [inline mne/realtime/tests/test_fieldtrip_client.py]: why is this necessary?

File: mne/realtime/tests/test_fieldtrip_client.py (hunk 17 above)
Comment (hekolk) [reply mne/realtime/tests/test_fieldtrip_client.py]: The numbers of epochs match. I inspected arrays in the debugger, some channels of the last epoch are incorrect. 

```
> percentage_of_correct_samples_per_ch = np.mean(np.isclose(rt_epochs.get_data()[-1, :, :],
                                                            epochs_offline.get_data()[-1, :, :]), 
                                                 axis=1)
> np.mean(percentage_of_correct_samples_per_ch[:314])
1.0
> np.mean(percentage_of_correct_samples_per_ch[314:])
0.03041912497126657
```
So somewhere the ``FieldTrip`` code returns invalid/incorrect buffer data (but not zeros)... Do you have a good prior where to look? It might as well be in the ``neuromag2ft`` code...

Comment (jasmainak) [reply mne/realtime/tests/test_fieldtrip_client.py]: okay that smells like a bug to me. I doubt it is in the `neuromag2ft` code though ... could it be some problem here: https://github.com/mne-tools/mne-python/blob/master/mne/realtime/fieldtrip_client.py#L350-L353 ?

Comment (hekolk) [reply mne/realtime/tests/test_fieldtrip_client.py]: Nevermind, the bug was an off-by-one error, since ``externals.FieldTrip.Client.getData`` returns the data inclusive the ``stop`` index. I'm not sure though why this wasn't caught in ``RtEpochs`` due to the buffer being too short for event processing...

File: mne/realtime/tests/test_fieldtrip_client.py (hunk 18 above)
Comment (hekolk) [reply mne/realtime/tests/test_fieldtrip_client.py]: to clean up the ``neuromag2ft`` process in ``finally`` (instead of relying on the timeout)

**Comment by hekolk** [comment]
>> Can we make this faster by changing the ISI, epoch length, and/or raw duration (e.g., via .crop) so it is more like 5 sec (without compromising the test quality)?

> There should be an option in neuromag2ft called --speed to make the simulation faster. Does that help?

The test takes ``~5s`` now (shorter raw duration and higher playback speed)

**Comment by hekolk** [comment]
> Travis can be slower compared to a desktop.
The test fails half of times on Travis despite higher timeouts; and always on the second test (the old one that is unchanged). My guess is that the previous ``neuromag2ft`` is still running and occupies the port (I tested locally that it doesn't fail if the port is taken).
I changed the code to using variable ports and checking them first - I'll trigger Travis a few times to see if the error persists.

**Comment by hekolk** [comment]
The test passed four times on Travis (builds 16376, 16378, 16384, 16387), so using dynamic ports fixed the issue.

**Comment by jasmainak** [comment]
This is great! I have one last nitpick though.

When I run the example `ftclient_rt_average.py` and try to do `epochs.plot()` or `epochs.save('abc.fif')` it does not work. It works fine for the other example though. I suppose the aim of this PR was to support saving the `RtEpochs`, right?

**Comment by hekolk** [comment]
Thanks for the scrutiny! :) I should have added tests for saving also for the other clients.
So it seems to be caused by calling ``RtEpochs.copy()`` while a ``FieldTripClient`` is attached, which fails on trying to pickle a ``Thread`` - so the implementation details of the client is the problem here.

1. add a ``__deepcopy__`` on ``RtEpochs`` that does not copy the ``_client``. → should work but violates the semantics of ``deepcopy`` (also the object would still not be pickable, but I think that would be fine).
2. make the client implementation responsible for being pickable/copyable (i.e., ignoring threads and sockets) → would not really solve the problem that we don't want a copy of a ``FieldTripClient`` that establishes a second network connection etc.
3. remove the ``_client`` reference from the ``RtEpochs`` state. The reference is only needed to get the ``Info``, register callbacks and start the receive thread (while a connection needs to have been started before to get an ``Info`` object). Since the user code creates a client anyway, it would be more natural to have an API similar to
```
client = SomeClientImplementation()
# optionally connect or use as context manager
rt_epochs = RtEpochs(client.get_info(), ...)
client.register_receiver(rt_epochs)
client.start_receive_thread()
``` 
→ while this seems to be a more sensible approach, it requires most changes and breaks/hinders backwards compability (?)

What are your thoughts on this? ... maybe I'm also missing the forest for the trees...

**Comment by larsoner** [comment]
I am fine with the current approach, it's simple enough to cover the corner case.

File: mne/realtime/tests/test_mockclient.py [hunk 19]
```diff
@@ -1,12 +1,19 @@
...
 
-from nose.tools import assert_true
-from numpy.testing import assert_array_equal
+import numpy as np
+from numpy.testing import assert_array_equal, assert_allclose
+import pytest
+
+# Set our plotters to test mode
+import matplotlib
+matplotlib.use('Agg')  # for testing don't use X server
```
Comment (larsoner) [inline mne/realtime/tests/test_mockclient.py]: Do this after other imports, this is causing Flake8 errors

File: mne/realtime/tests/test_fieldtrip_client.py [hunk 20]
```diff
@@ -8,12 +8,20 @@
...
+import socket
 
-from nose.tools import assert_true, assert_equal
+import numpy as np
+from numpy.testing import assert_array_equal, assert_allclose
+import pytest
+
+# Set our plotters to test mode
+import matplotlib
+matplotlib.use('Agg')  # for testing don't use X server
```
Comment (larsoner) [inline mne/realtime/tests/test_fieldtrip_client.py]: same

**Comment by jasmainak** [comment]
Let me know when I should test again :)

**Comment by hekolk** [comment]
I used the occasion to also add some tests for the other public methods of ``BaseEpochs`` (mostly smoke tests).

@jasmainak feel free to have another look

File: mne/realtime/tests/test_fieldtrip_client.py [hunk 21]
```diff
@@ -8,70 +8,155 @@
...
-from nose.tools import assert_true, assert_equal
+from .test_mockclient import _call_base_epochs_public_api
 
-import mne
-from mne.utils import requires_neuromag2ft, run_tests_if_main
-from mne.realtime import FieldTripClient
-from mne.externals.six.moves import queue
+# Set our plotters to test mode
+import matplotlib
+matplotlib.use('Agg')  # for testing don't use X server
```
Comment (jasmainak) [inline mne/realtime/tests/test_fieldtrip_client.py]: shouldn't this be imported before everything else? Because mne also imports matplotlib

**Comment by jasmainak** [comment]
okay LGTM

let's merge @larsoner ?

**Comment by larsoner** [comment]
Thanks @hekolk !
